targets/psa_switch/externs/psa_hash.cpp

```diff
@@ -0,0 +1,118 @@
+/* Copyright 2021 SYRMIA LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/*
+ * Dusan Krdzic (dusan.krdzic@syrmia.com)
+ *
+ */
+
+
+#include "psa_hash.h"
+
+namespace {
+
+std::string
+prepare_data_for_algo(const std::vector<bm::Field> &fields) {
```

Comment:
I know about this approach and have already tried it this way. However, it does not work for fields whose length is indivisible by 8. e.g. bit<12> a = 0x456 will give hash `0xb832` which is not correct. It should be `0xfe82`. 
If you think that field length must be divisible with 8, I will push new implementation with `BufBuilder` help class

Reply:
`BufBuilder` pads to right. It's possible that CRCs expects padding to the left (at least that's what I see here: https://stackoverflow.com/questions/3411654/best-way-to-generate-crc8-16-when-input-is-odd-number-of-bits-not-byte-c-or-p). But what about non-CRC hash algorithms? I guess padding to the left is reasonable, so I am fine with modifying the implementation to do that. As long as `field.deparse` is used to avoid big numbers or custom code it works for me.

That may mean iterating over the field list twice: once to compute the required length and once to deparse (serialize) fields. But that's probably not a big issue, as it also lets us allocate the required memory ahead of time. Hope that works:

```cpp
int nbits = 0;
for (const auto &field : fields) {
  nbits += field.get_nbits();
}
int nbytes = (nbits + 7) / 8;
ByteContainer buf(nbytes); // 0-initialized
char *ptr = buf->data();
nbits = (nbytes * 8 - nbits);  // pad to the left with 0s
for (const auto &field : fields) {
  ptr += (nbits + 7) / 8;
  field.deparse(ptr, nbits % 8);
  nbits += field.get_nbits();
}
compute(buf.data(), buf.size());
```

Reply:
I just tried this approach. It doesn't seem to work. It calculates bad hash again now even for bit size divisible by 8. I’m not sure that 2 `Data` vars could degrade performance that much. They make code simpler, readable and it works what is important.

Reply:
I modified this approach a bit and now it works.